common/ssh_max_arg.py

```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env python3
+# SPDX-FileCopyrightText: © 2015-2022 Germar Reitze
+#
+# SPDX-License-Identifier: GPL-2.0-or-later
+#
+# This file is part of the program "Back In time" which is released under GNU
+# General Public License v2 (GPLv2). See file/folder LICENSE or go to
+# <https://spdx.org/licenses/GPL-2.0-or-later.html>.
+"""This module determines the maximum possible length of an SSH command.
+
+It can also can run as a stand alone script. The solution is based on
+https://www.theeggeadventure.com/wikimedia/index.php/Ssh_argument_length
+"""
+import random
+import string
+import subprocess
+import socket
+import argparse
+from config import Config
+
+# Must be divisible by 8
+_INITIAL_SSH_CMD_SIZE = 1048320
+_ERR_CODE_E2BIG = 7
+
+
+def probe_max_ssh_command_size(
+        cfg: Config,
+        ssh_command_size: int = _INITIAL_SSH_CMD_SIZE,
+        size_offset: int = _INITIAL_SSH_CMD_SIZE) -> int:
+    """Determine the maximum length of SSH commands for the current config
+
+    Try a SSH command with length ``ssh_command_size``. The command is
+    decreased by ``size_offset`` if it was too long or increased if it worked.
+    The function calls itself recursively until it finds the maximum
+    possible length. The offset ``size_offset`` is bisect in each try.
+
+    Args:
+        config Back In Time config instance including the details about the
+            current SSH snapshot profile. The current profile must use the SSH
+            mode.
+        ssh_command_size: Initial length used for the test argument.
+        size_offset: Offset for increase or decrease ``ssh_command_size``.
+
+    Returns:
+        The maximum possible SSH command length.
+
+    Raises:
+        Exception: If there are unhandled cases or the recurse ends in an
+                   undefined state.
+        OSError: If there are unhandled cases.
+    """
+    size_offset = round(size_offset / 2)
+
+    # random string of desired length
+    command_string = ''.join(random.choices(
+        string.ascii_uppercase+string.digits, k=ssh_command_size))
+
+    # use that string in a printf statement via SSH
+    ssh = cfg.sshCommand(
+        cmd=['printf', command_string],
+        nice=False,
+        ionice=False,
+        prefix=False)
+
+    try:
+        with subprocess.Popen(ssh,
+                              stdout=subprocess.PIPE,
+                              stderr=subprocess.PIPE,
+                              universal_newlines=True) as proc:
+            out, err = proc.communicate()
```

Comment:
Consider using `subprocess.run`:

```suggestion
        proc = subprocess.run(capture_output=True, text=True)
        out, err = proc.stdout, proc.stderr
```

Details from [subprocess docs](https://docs.python.org/3/library/subprocess.html):

- The recommended approach to invoking subprocesses is to use the [run()](https://docs.python.org/3/library/subprocess.html#subprocess.run) function for all use cases it can handle.
- If _capture_output_ is true, stdout and stderr will be captured. When used, the internal [Popen](https://docs.python.org/3/library/subprocess.html#subprocess.Popen) object is automatically created with stdout and stderr both set to [PIPE](https://docs.python.org/3/library/subprocess.html#subprocess.PIPE).
- If ... _text_ is true, file objects for stdin, stdout and stderr are opened in text mode using the specified encoding and errors or the [io.TextIOWrapper](https://docs.python.org/3/library/io.html#io.TextIOWrapper) default. The _universal_newlines_ argument is equivalent to _text_ and is provided for backwards compatibility.